mmaprototype: add set_store test and allow attrs/localities to  change

tbg opened the pull request:
This updates cs.setStore so that it reacts correctly to changed store/node attributes or locality tiers by updating its internal state accordingly.

Closes https://github.com/cockroachdb/cockroach/issues/157831.

Epic: CRDB-55052

tbg committed:
mmaprototype: add set_store test

This documents that updates store attributes and localities are ignored.

tbg committed:
mmaprototype: allow locality and attrs to change

This can occur if a non-local store restarts and comes back with changed
locality or (store or node) attributes.

Informs #157757.

Epic: CRDB-55052

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/157833)
<!-- Reviewable:end -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157833)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store` line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2B2Wm0MHFz4rE46xN:-Oe2B2Wm0MHFz4rE46xO:bamznzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store#L51)):*
> ```
> node-id=2 locality-tiers=region=us-two,node=2
>   store-id=2 attrs=twotwo locality-code=6:7:
> ```

Highlighting.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 2292 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2B15N9Yl7xWpRkCCB:-Oe2B15N9Yl7xWpRkCCC:bamznzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L2292)):*
> ```Go
> 		cs.stores[sal.StoreID].localityTiers = cs.localityTierInterner.intern(sal.locality())
> 		cs.stores[sal.StoreID].StoreAttributesAndLocality = sal
> 		// TODO(during review): is there more that needs to be updated when the
> ```

Highlighting.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> @sumeerbhola reviewed 4 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157833)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 2292 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2B15N9Yl7xWpRkCCB:-Oe2FsY3Ei3_g0Sp3vD3:b-6bsyw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L2292)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Highlighting.
</blockquote></details>

This is sufficient.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 2278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2EJIkDI4pJNxJ4YxQ:-Oe2EJIkDI4pJNxJ4YxR:bxo8e99) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L2278)):*
> ```Go
> 	// If the store is new or the locality/attributes changed, we need to update
> 	// the locality and attributes.
> 	loc := sal.locality()
> ```

This results in a slice allocation.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2HHdD4xLqBsNi28wQ:-Oe2HHdD4xLqBsNi28wR:b36plyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store#L5)):*
> ```
> ----
> node-id=1 locality-tiers=region=europe,node=1
>   store-id=1 attrs=orange,black locality-code=1:2:
> ```

in addition to the locality-code, I think we should use `localityTierInterner.unintern` and print the human readable string. Or we could just also assert inside the test code it is the same as what is returned by `StoreAttributesAndLocality.locality`, to spare the testdata reviewer from the cognitive burden of checking equality in this testdata file.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2GoqJ31wIyP4aNUel:-Oe2GoqJ31wIyP4aNUem:b-91xghz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store#L25)):*
> ```
>   store-id=1 node-id=1 node-attrs=hicpu,himem locality-tiers=planet=earth,island=kiribati,node=1
> ----
> node-id=1 locality-tiers=planet=earth,island=kiribati,node=1,node=1 node-attrs=hicpu,himem
> ```

was this `node=1,node=1` a deliberate choice to test?
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store` line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2DoD50XDMHIaHthgp:-Oe2DoD50XDMHIaHthgq:b2raf0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store#L53)):*
> ```
> 
> # TODO(during review): we should also be able to print the state of the constraintMatcher
> # here. I only noticed at the end that I wasn't updating the constraintMatcher on changes,
> ```

is this TODO for a later PR?


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-:-OeqEdcm4CbLRHm80V_r:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)



<!-- Sent from Reviewable.io -->

tbg committed:
mmaprototype: introduce storeAttributesAndLocalityWithNodeTier

This removes a potential footgun. Previously, internal to mma,
we always wanted to use StoreAttributesAndLocality.locality()
which has an extra tier for the node. Now, internal to mma,
we just never use StoreAttributesAndLocality directly, so
there's no risk of accidentally forgetting this call - we
now use the NodeLocality field directly in all places.

tbg reviewed (commented):
Added a round of updates for Sumeer's comments and took care of the TODOs.
It's worth a second look.

I did amend and force-push (as well as add a new commit). Reviewable should be handling this well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-:-OeqIqaz7hH3JkNyoJOl:b-gf2j91)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 2278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2EJIkDI4pJNxJ4YxQ:-OeqEft5CxbIbfBOYECP:bakejt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L2278)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This results in a slice allocation.
</blockquote></details>

I'm not sure this allocation matters (this is on the incoming gossip handler path, so not ultra frequent and already correlated with a handful of allocs at least in practice), but at least we shouldn't be calling `locality()` twice (as this code was).

I added a cleanup commit here: I didn't like that we needed to remember to call `locality()` instead of using the `NodeLocality` field. I added an mma-local type that represents "sal but with the tier added already". This seems cleaner. We still have to add the extra tier once, at the caller, though. If you think this should be optimized further (I don't think so) I can probably manage somehow, though it'll require some more re-working.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 2292 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2B15N9Yl7xWpRkCCB:-OeqImKF-xtumMvWm0Pp:b-eem509) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L2292)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This is sufficient.
</blockquote></details>

Ack, removed comment.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2HHdD4xLqBsNi28wQ:-OeofaZo-ZfHZ0rs5wiS:b-clx6p6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store#L5)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

in addition to the locality-code, I think we should use `localityTierInterner.unintern` and print the human readable string. Or we could just also assert inside the test code it is the same as what is returned by `StoreAttributesAndLocality.locality`, to spare the testdata reviewer from the cognitive burden of checking equality in this testdata file.
</blockquote></details>

Good idea, printing the human-readable string now and asserting that there's no drift between what's interned and what's stored in `StoreAttributesAndLocality`.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2GoqJ31wIyP4aNUel:-OeofnYL5HhuADYbaeP_:bacakhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store#L25)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

was this `node=1,node=1` a deliberate choice to test?
</blockquote></details>

No, this was copy pasta. Fixed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store` line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2DoD50XDMHIaHthgp:-OeogqRZAXDqUXneTz1D:b5vtszc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/set_store#L53)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this TODO for a later PR?
</blockquote></details>

No, it's for here.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
(if it doesn't lmk and I'll put up a diffed version)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-:-OeqJY_m3VuplPighWDd:bcxakmi)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola and @wenyihu6)



<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 15 of 19 files at r4, 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-:-OeqfkjZ1eGVq__S3_rf:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 2278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157833#-Oe2EJIkDI4pJNxJ4YxQ:-OeqerCZ7tTi7ZWoPbdE:b2nhb2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8522f60675b513baee78695950ad6139b3769e/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L2278)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'm not sure this allocation matters (this is on the incoming gossip handler path, so not ultra frequent and already correlated with a handful of allocs at least in practice), but at least we shouldn't be calling `locality()` twice (as this code was).

I added a cleanup commit here: I didn't like that we needed to remember to call `locality()` instead of using the `NodeLocality` field. I added an mma-local type that represents "sal but with the tier added already". This seems cleaner. We still have to add the extra tier once, at the caller, though. If you think this should be optimized further (I don't think so) I can probably manage somehow, though it'll require some more re-working.
</blockquote></details>

Ack


<!-- Sent from Reviewable.io -->
